File: api/export_excel.js
const ExcelJS = require('exceljs');
const path = require('path');
const fs = require('fs');

const TEMPLATE_PATH = path.join(__dirname, '..', 'template.xlsx');

async function fillPlan(data, wb) {
  const ws = wb.getWorksheet(1);
  const ui       = data.userInfo   || {};
  const goals    = data.goals      || {};
  const services = data.weeklyServices || {};
  const created  = data.createdAt  || '';

  const w = (cell, value, size = 8) => {
    const c = ws.getCell(cell);
    c.value = value;
    c.font = { ...(c.font || {}), size, name: 'メイリオ' };
  };
  const wa = (cell, value, size = 8) => {
    const c = ws.getCell(cell);
    c.value = value;
    c.font = { size, name: 'メイリオ' };
    c.alignment = { wrapText: true, vertical: 'top' };
  };

  // 作成日
  let dt;
  try { dt = created ? new Date(created) : new Date(); } catch { dt = new Date(); }
  w('H1', dt.getFullYear(), 9); w('K1', dt.getMonth()+1, 9); w('M1', dt.getDate(), 9);
  w('H2', ui.svcresp || '', 9);

  // 利用者情報
  w('E3', ui.name || '', 10);
  w('N3', ui.gender === '男' ? '■男　　□ 女' : '□男　　■ 女', 9);
  const birth = ui.birth || '';
  w('V3', birth.includes('昭和') ? '□明治　□大正　■昭和' : '□明治　□大正　□昭和', 8);
  const bm = birth.match(/(\d+)年\s*(\d+)月\s*(\d+)日/);
  if (bm) { w('AB3', +bm[1]); w('AD3', +bm[2]); w('AF3', +bm[3]); }
  w('AG3', ui.addr || '', 8);

  // 緊急連絡先
  w('E4',  ui.emg1_name || '', 9);
  w('L4',  '続柄：' + (ui.emg1_rel || ''), 8);
  w('S4',  ui.emg1_tel  || '', 8);
  w('AD4', ui.emg2_name || '', 9);
  w('AK4', '続柄：' + (ui.emg2_rel || ''), 8);
  w('AR4', ui.emg2_tel  || '', 8);

  // 事業所
  w('E5',  ui.office      || '', 9);
  w('O5',  ui.office_tel  || '', 8);
  w('V5',  ui.office_addr || '', 8);
  w('AM5', ui.manager     || '', 9);
  w('E6',  ui.svcresp     || '', 9);
  w('T6',  ui.helper      || '', 9);
  w('L7',  (ui.cm_name||'') + '　' + (ui.cm_office||''), 8);
  w('Z7',  ui.cm_tel    || '', 8);
  w('AH7', ui.doctor    || '', 8);
  w('AR7', ui.doctor_tel|| '', 8);

  // 計画期間
  const period = ui.period || '';
  const pm = [...period.matchAll(/(\d{4})年(\d{1,2})月(\d{1,2})日/g)];
  if (pm.length >= 2) {
    w('H9',+pm[0][1]); w('J9',+pm[0][2]); w('L9',+pm[0][3]);
    w('O9',+pm[1][1]); w('Q9',+pm[1][2]); w('S9',+pm[1][3]);
  }

  // 週間サービス
  const dayStartCol = { 月:3, 火:8, 水:13, 木:18, 金:23, 土:28, 日:33 };
  const timeRows = {6:12,7:14,8:16,9:18,10:20,11:22,12:24,13:26,14:28,15:30,16:32,17:34,18:36,19:38,20:40,21:42,22:44,23:46};
  let bodyRow = timeRows[9];
  let lifeRow = timeRows[14];

  const colLetter = n => {
    let s = '';
    while (n > 0) { s = String.fromCharCode(64 + (n % 26 || 26)) + s; n = Math.floor((n-1)/26); }
    return s;
  };

  for (const [svcName, svcInfo] of Object.entries(services)) {
    const svcType = svcInfo.type || '';
    const svcDays = svcInfo.days || [];
    const row  = svcType === '身体介護' ? bodyRow : lifeRow;
    const icon = svcType === '身体介護' ? '▶' : '◆';
    const label = `${icon}${svcName}`;

    for (const day of svcDays) {
      const sc = dayStartCol[day];
      if (!sc) continue;
      const addr = `${colLetter(sc)}${row}`;
      const c = ws.getCell(addr);
      c.value = c.value ? `${c.value}\n${label}` : label;
      c.font = { size: 7, name: 'メイリオ' };
      c.alignment = { wrapText: true, vertical: 'top' };
    }

    if (svcType === '身体介護') bodyRow = Math.min(bodyRow + 2, lifeRow - 2);
    else lifeRow = Math.min(lifeRow + 2, 46);
  }

  // 援助目標・留意事項
  const notesList = Array.isArray(goals.notes) ? goals.notes : [String(goals.notes||'')];
  const notesStr = notesList.map(n => `・${n}`).join('\n');
  const goalText = `【長期目標】\n${goals.long||''}\n\n【短期目標①】\n${goals.short1||''}\n\n【短期目標②】\n${goals.short2||''}\n\n【留意事項】\n${notesStr}`;
  wa('AM10', goalText, 8);
  ws.getRow(10).height = 200;
  for (let r = 11; r < 48; r++) {
    const row = ws.getRow(r);
    if (!row.height || row.height < 25) row.height = 25;
  }
}

module.exports = async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method Not Allowed' });

  try {
    const data = req.body;

    const wb = new ExcelJS.Workbook();
    await wb.xlsx.readFile(TEMPLATE_PATH);
    await fillPlan(data, wb);

    const buf = await wb.xlsx.writeBuffer();

    res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
    res.setHeader('Content-Disposition', "attachment; filename*=UTF-8''%E8%A8%AA%E5%95%8F%E4%BB%8B%E8%AD%B7%E8%A8%88%E7%94%BB%E6%9B%B8.xlsx");
    res.status(200).send(Buffer.from(buf));
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: err.message });
  }
};

File: api/export_excel.py
import json, re, io, os
from datetime import datetime
from http.server import BaseHTTPRequestHandler
from openpyxl import load_workbook
from openpyxl.styles import Font, Alignment
from openpyxl.utils import get_column_letter

TEMPLATE_PATH = os.path.join(os.path.dirname(__file__), "..", "template.xlsx")

def fill_plan(data: dict, wb) -> None:
    ws = wb.active
    ui       = data.get("userInfo", {})
    goals    = data.get("goals", {})
    services = data.get("weeklyServices", {})
    created  = data.get("createdAt", "")

    def w(cell, value, size=8):
        c = ws[cell]; c.value = value
        c.font = Font(size=size, name="メイリオ")

    def wa(cell, value, size=8):
        c = ws[cell]; c.value = value
        c.font = Font(size=size, name="メイリオ")
        c.alignment = Alignment(wrap_text=True, vertical="top")

    try:
        dt = datetime.fromisoformat(created[:10]) if created else datetime.today()
    except:
        dt = datetime.today()

    w("H1", dt.year, 9);  w("K1", dt.month, 9); w("M1", dt.day, 9)
    w("H2", ui.get("svcresp", ""), 9)
    w("E3", ui.get("name", ""), 10)
    gender = ui.get("gender", "")
    w("N3", "■男　　□ 女" if gender == "男" else "□男　　■ 女", 9)
    birth = ui.get("birth", "")
    w("V3", "□明治　□大正　■昭和" if "昭和" in birth else "□明治　□大正　□昭和", 8)
    m = re.search(r'(\d+)年\s*(\d+)月\s*(\d+)日', birth)
    if m:
        w("AB3", int(m.group(1))); w("AD3", int(m.group(2))); w("AF3", int(m.group(3)))
    w("AG3", ui.get("addr", ""), 8)
    w("E4",  ui.get("emg1_name", ""), 9)
    w("L4",  "続柄：" + ui.get("emg1_rel", ""), 8)
    w("S4",  ui.get("emg1_tel", ""), 8)
    w("AD4", ui.get("emg2_name", ""), 9)
    w("AK4", "続柄：" + ui.get("emg2_rel", ""), 8)
    w("AR4", ui.get("emg2_tel", ""), 8)
    w("E5",  ui.get("office", ""), 9)
    w("O5",  ui.get("office_tel", ""), 8)
    w("V5",  ui.get("office_addr", ""), 8)
    w("AM5", ui.get("manager", ""), 9)
    w("E6",  ui.get("svcresp", ""), 9)
    w("T6",  ui.get("helper", ""), 9)
    w("L7",  ui.get("cm_name","") + "　" + ui.get("cm_office",""), 8)
    w("Z7",  ui.get("cm_tel", ""), 8)
    w("AH7", ui.get("doctor", ""), 8)
    w("AR7", ui.get("doctor_tel", ""), 8)

    period = ui.get("period", "")
    pm = re.findall(r'(\d{4})年(\d{1,2})月(\d{1,2})日', period)
    if len(pm) >= 2:
        w("H9",int(pm[0][0])); w("J9",int(pm[0][1])); w("L9",int(pm[0][2]))
        w("O9",int(pm[1][0])); w("Q9",int(pm[1][1])); w("S9",int(pm[1][2]))

    day_start_col = {"月":3,"火":8,"水":13,"木":18,"金":23,"土":28,"日":33}
    time_rows = {6:12,7:14,8:16,9:18,10:20,11:22,12:24,13:26,14:28,
                 15:30,16:32,17:34,18:36,19:38,20:40,21:42,22:44,23:46}
    body_row = time_rows[9]
    life_row = time_rows[14]

    def get_parent(addr):
        for mr in ws.merged_cells.ranges:
            cells = [f"{get_column_letter(cc)}{rr}"
                     for rr in range(mr.min_row, mr.max_row+1)
                     for cc in range(mr.min_col, mr.max_col+1)]
            if addr in cells:
                return f"{get_column_letter(mr.min_col)}{mr.min_row}"
        return addr

    for svc_name, svc_info in services.items():
        svc_type = svc_info.get("type","")
        svc_days = svc_info.get("days",[])
        row  = body_row if svc_type == "身体介護" else life_row
        icon = "▶" if svc_type == "身体介護" else "◆"
        label = f"{icon}{svc_name}"
        for day in svc_days:
            sc = day_start_col.get(day)
            if sc is None: continue
            addr = get_parent(f"{get_column_letter(sc)}{row}")
            existing = ws[addr].value
            ws[addr].value = (str(existing) + "\n" + label) if existing else label
            ws[addr].font = Font(size=7, name="メイリオ")
            ws[addr].alignment = Alignment(wrap_text=True, vertical="top")
        if svc_type == "身体介護":
            body_row = min(body_row + 2, life_row - 2)
        else:
            life_row = min(life_row + 2, 46)

    notes_list = goals.get("notes", [])
    notes_str = "\n".join(f"・{n}" for n in (notes_list if isinstance(notes_list, list) else [str(notes_list)]))
    goal_text = (
        f"【長期目標】\n{goals.get('long','')}\n\n"
        f"【短期目標①】\n{goals.get('short1','')}\n\n"
        f"【短期目標②】\n{goals.get('short2','')}\n\n"
        f"【留意事項】\n{notes_str}"
    )
    wa("AM10", goal_text, 8)
    ws.row_dimensions[10].height = 200
    for r in range(11, 48):
        if ws.row_dimensions[r].height is None or ws.row_dimensions[r].height < 25:
            ws.row_dimensions[r].height = 25


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200); self._cors(); self.end_headers()

    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        try:
            data = json.loads(self.rfile.read(length))
        except Exception as e:
            self._json(400, {"error": str(e)}); return

        try:
            wb = load_workbook(TEMPLATE_PATH)
            fill_plan(data, wb)
            out = io.BytesIO()
            wb.save(out)
            xlsx = out.getvalue()
        except Exception as e:
            self._json(500, {"error": str(e)}); return

        self.send_response(200); self._cors()
        self.send_header("Content-Type", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
        self.send_header("Content-Disposition", "attachment; filename*=UTF-8''%E8%A8%AA%E5%95%8F%E4%BB%8B%E8%AD%B7%E8%A8%88%E7%94%BB%E6%9B%B8.xlsx")
        self.send_header("Content-Length", str(len(xlsx)))
        self.end_headers(); self.wfile.write(xlsx)

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def _json(self, status, obj):
        body = json.dumps(obj).encode()
        self.send_response(status); self._cors()
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers(); self.wfile.write(body)

    def log_message(self, *args): pass
